diff --git a/deps/v8/src/api/api.cc b/deps/v8/src/api/api.cc
index 2dd476dda34..cc1b8d41680 100644
--- a/deps/v8/src/api/api.cc
+++ b/deps/v8/src/api/api.cc
@@ -11268,6 +11268,15 @@ void HeapSnapshot::Serialize(OutputStream* stream,
   serializer.Serialize(stream);
 }
 
+RedactedHeapSnapshot::RedactedHeapSnapshot(const HeapSnapshot* snapshot):
+  snapshot_(snapshot) {
+}
+
+void RedactedHeapSnapshot::Serialize(OutputStream* stream) const {
+  i::HeapSnapshotJSONSerializer serializer(ToInternal(snapshot_));
+  serializer.Serialize(stream, true);
+}
+
 // static
 STATIC_CONST_MEMBER_DEFINITION const SnapshotObjectId
     HeapProfiler::kUnknownObjectId;
